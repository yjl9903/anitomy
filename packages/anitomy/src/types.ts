import type { ElementCategory } from './element';

export interface AnitomyOptions {
  delimiters: string;

  parseEpisodeNumber: boolean;

  parseEpisodeTitle: boolean;

  parseFileExtension: boolean;

  parseReleaseGroup: boolean;
}

export type ParsedResult = Partial<Record<ElementCategory, string>>;

export interface AnitomyResult {
  title: string | undefined;

  type: string | undefined;

  season: string | undefined;

  year: number | undefined;

  month: number | undefined;

  language: string | undefined;

  subtitles: string | undefined;

  source: string | undefined;

  episode: {
    number: number | undefined;

    numberAlt: number | undefined;

    title: string | undefined;
  };

  volume: {
    number: number | undefined;
  };

  video: {
    term: string | undefined;

    resolution: string | undefined;
  };

  audio: {
    term: string | undefined;
  };

  release: {
    version: number | undefined;
    group: string | undefined;
  };

  file: {
    name: string;

    checksum: string | undefined;

    extension: string | undefined;
  };

  // prefix: {
  //   season: string | undefined;
  //   volume: string | undefined;
  //   episode: string | undefined;
  // };
}
