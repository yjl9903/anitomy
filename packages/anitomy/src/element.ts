export enum ElementCategory {
  AnimeSeason = 'season',
  AnimeSeasonPrefix = 'prefix.season',
  AnimeTitle = 'title',
  AnimeType = 'type',
  AnimeYear = 'year',
  AnimeMonth = 'month', // Added by https://github.com/yjl9903/AnimeGarden
  DeviceCompatibility = 'DeviceCompatibility',
  Source = 'source',
  // Episode
  EpisodeNumber = 'episode.number',
  EpisodeNumberAlt = 'episode.numberAlt',
  EpisodePrefix = 'prefix.episode',
  EpisodeTitle = 'episode.title',
  // File
  FileChecksum = 'checksum',
  FileExtension = 'extension',
  FileName = 'filename',
  // Language
  Language = 'language',
  Subtitles = 'subtitles',
  // Audio and Video
  AudioTerm = 'audio.term',
  VideoResolution = 'video.resolution',
  VideoTerm = 'video.term',
  // Volume
  VolumeNumber = 'volume',
  VolumePrefix = 'prefix.volume',
  // Release
  ReleaseGroup = 'release.group',
  ReleaseInformation = 'release.information',
  ReleaseVersion = 'release.version',
  // Other
  Unknown = 'unknown',
  Other = 'other'
}
