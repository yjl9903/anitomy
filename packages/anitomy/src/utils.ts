import type { ParsedResult } from './types';

export function inRange<T>(list: T[], idx: number) {
  return 0 <= idx && idx < list.length;
}

export function isNumericString(text: string) {
  return /^\d+$/.test(text);
}

export function trim(text: string, removal: string[]) {
  let start = 0,
    end = text.length - 1;
  while (start <= end && removal.includes(text[start])) {
    start++;
  }
  while (end >= start && removal.includes(text[end])) {
    end--;
  }
  return text.slice(start, end + 1);
}

export function mergeResult(source: ParsedResult, income: ParsedResult = {}) {
  return {
    ...source,
    ...income
  };
}
