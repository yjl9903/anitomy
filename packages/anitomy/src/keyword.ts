import type { ParsedResult } from './types';

import { TextRange } from './token';
import { ElementCategory } from './element';

export interface KeywordOptions {
  identifiable: boolean;

  searchable: boolean;

  valid: boolean;
}

export type Keyword = {
  category: ElementCategory;
} & KeywordOptions;

const keys = new Map<string, Keyword>();

const extensions = new Map<string, Keyword>();

export class KeywordManager {
  private static keys = keys;

  private static extensions = extensions;

  private static peekEntries: Array<{ category: ElementCategory; list: string[] }> = [
    { category: ElementCategory.AudioTerm, list: ['Dual Audio'] },
    { category: ElementCategory.VideoTerm, list: ['H264', 'H.264', 'h264', 'h.264'] },
    {
      category: ElementCategory.VideoResolution,
      list: ['480p', '480P', '720p', '720P', '1080p', '1080P']
    },
    { category: ElementCategory.Source, list: ['Blu-Ray'] }
  ];

  static {
    const optionsDefault: KeywordOptions = { identifiable: true, searchable: true, valid: true };
    const optionsInvalid: KeywordOptions = { identifiable: false, searchable: false, valid: false };
    const optionsUnidentifiable: KeywordOptions = {
      identifiable: false,
      searchable: true,
      valid: true
    };
    const optionsUnidentifiableInvalid: KeywordOptions = {
      identifiable: false,
      searchable: true,
      valid: false
    };
    const optionsUnidentifiableUnsearchable: KeywordOptions = {
      identifiable: false,
      searchable: false,
      valid: true
    };

    // Season
    add(ElementCategory.AnimeSeasonPrefix, optionsUnidentifiable, ['SAISON', 'SEASON']);

    // Type
    add(ElementCategory.AnimeType, optionsUnidentifiable, [
      'GEKIJOUBAN',
      'MOVIE',
      'OAD',
      'OAV',
      'ONA',
      'OVA',
      'SPECIAL',
      'SPECIALS',
      'TV',
      '特别篇',
      '特別篇',
      '特別編',
      '特别话',
      '特別话',
      '特別話',
      '番外篇',
      '番外編'
    ]);
    add(ElementCategory.AnimeType, optionsUnidentifiableUnsearchable, ['SP']);
    add(ElementCategory.AnimeType, optionsUnidentifiableInvalid, [
      'ED',
      'ENDING',
      'NCED',
      'NCOP',
      'OP',
      'OPENING',
      'PREVIEW',
      'PV'
    ]);

    add(ElementCategory.AudioTerm, optionsDefault, [
      // Audio channels
      '2.0CH',
      '2CH',
      '5.1',
      '5.1CH',
      'DTS',
      'DTS-ES',
      'DTS5.1',
      'TRUEHD5.1',
      // Audio codec
      'AAC',
      'AACX2',
      'AACX3',
      'AACX4',
      'AC3',
      'EAC3',
      'E-AC-3',
      'FLAC',
      'FLACX2',
      'FLACX3',
      'FLACX4',
      'LOSSLESS',
      'MP3',
      'OGG',
      'VORBIS',
      // Audio language
      'DUALAUDIO',
      'DUAL AUDIO'
    ]);
    add(ElementCategory.VideoTerm, optionsDefault, [
      // Frame rate
      '23.976FPS',
      '24FPS',
      '29.97FPS',
      '30FPS',
      '60FPS',
      '120FPS',
      // Video codec
      '8BIT',
      '8-BIT',
      '10BIT',
      '10BITS',
      '10-BIT',
      '10-BITS',
      'HI10',
      'HI10P',
      'HI444',
      'HI444P',
      'HI444PP',
      'H264',
      'H265',
      'H.264',
      'H.265',
      'X264',
      'X265',
      'X.264',
      'AVC',
      'HEVC',
      'HEVC2',
      'HEVC-10BIT',
      'DIVX',
      'DIVX5',
      'DIVX6',
      'XVID',
      // Video format
      'AVI',
      'RMVB',
      'WMV',
      'WMV3',
      'WMV9',
      // Video quality
      'HQ',
      'LQ',
      // Video resolution
      'HD',
      'SD'
    ]);

    add(ElementCategory.Source, optionsDefault, [
      'BD',
      'BDRIP',
      'BLURAY',
      'BLU-RAY',
      'DVD',
      'DVD5',
      'DVD9',
      'DVD-R2J',
      'DVDRIP',
      'DVD-RIP',
      'R2DVD',
      'R2J',
      'R2JDVD',
      'R2JDVDRIP',
      'HDTV',
      'HDTVRIP',
      'TVRIP',
      'TV-RIP',
      'WEBCAST',
      'WEBDL',
      'WEB-DL',
      'WEBRIP',
      'WEB-RIP'
    ]);

    // Language
    add(ElementCategory.Language, optionsDefault, [
      'ENG',
      'ENGLISH',
      'ESPANO',
      'JAP',
      'PT-BR',
      'SPANISH',
      'VOSTFR',
      // feat: chinese
      'CHT',
      'CHS',
      '简中',
      '繁中',
      '简体',
      '繁體'
    ]);
    add(ElementCategory.Language, optionsUnidentifiable, ['ESP', 'ITA']);

    add(ElementCategory.Subtitles, optionsDefault, [
      'ASS',
      'GB', // feat: 简体字幕
      'BIG5',
      'DUB',
      'DUBBED',
      'HARDSUB',
      'HARDSUBS',
      'RAW',
      'SOFTSUB',
      'SOFTSUBS',
      'SUB',
      'SUBBED',
      'SUBTITLED'
    ]);

    // Episode
    add(ElementCategory.EpisodePrefix, optionsDefault, [
      'EP',
      'EP.',
      'EPS',
      'EPS.',
      'EPISODE',
      'EPISODE.',
      'EPISODES',
      'CAPITULO',
      'EPISODIO',
      'FOLGE'
    ]);
    add(ElementCategory.EpisodePrefix, optionsInvalid, ['E', '\\x7B2C']);

    // Volume
    add(ElementCategory.VolumePrefix, optionsDefault, ['VOL', 'VOL.', 'VOLUME']);

    // Release
    add(ElementCategory.ReleaseGroup, optionsDefault, ['Baha', 'THORA']);
    add(ElementCategory.ReleaseInformation, optionsDefault, [
      'BATCH',
      'COMPLETE',
      'PATCH',
      'REMUX'
    ]);
    add(ElementCategory.ReleaseInformation, optionsUnidentifiable, ['END', 'FINAL']);
    add(ElementCategory.ReleaseVersion, optionsDefault, [
      'V0',
      'V1',
      'V2',
      'V3',
      'V4',
      'V5',
      'V6',
      'V7',
      'V8',
      'V9'
    ]);

    // Other
    add(ElementCategory.Other, optionsDefault, [
      'REMASTER',
      'REMASTERED',
      'UNCENSORED',
      'UNCUT',
      'TS',
      'VFR',
      'WIDESCREEN',
      'WS'
    ]);

    // File extension
    add(ElementCategory.FileExtension, optionsDefault, [
      '3GP',
      'AVI',
      'DIVX',
      'FLV',
      'M2TS',
      'MKV',
      'MOV',
      'MP4',
      'MPG',
      'OGM',
      'RM',
      'RMVB',
      'TS',
      'WEBM',
      'WMV'
    ]);
    add(ElementCategory.FileExtension, optionsInvalid, [
      'AAC',
      'AIFF',
      'FLAC',
      'M4A',
      'MP3',
      'MKA',
      'OGG',
      'WAV',
      'WMA',
      '7Z',
      'RAR',
      'ZIP',
      'ASS',
      'SRT'
    ]);

    function add(category: ElementCategory, options: KeywordOptions, input: string[]) {
      const map = category === 'extension' ? extensions : keys;
      for (const key of input) {
        if (!map.has(key)) {
          map.set(key, { category, ...options });
        }
      }
    }
  }

  public static normalize(text: string) {
    return text.toLocaleUpperCase();
  }

  public static contains(category: ElementCategory, keyword: string) {
    const map = this.container(category);
    const value = map.get(keyword);
    return value && value.category === category;
  }

  public static container(category: ElementCategory) {
    return category === 'extension' ? this.extensions : this.keys;
  }

  public static find(keyword: string, category: ElementCategory) {
    const map = this.container(category);
    if (!map.has(keyword)) {
      return undefined;
    }
    const entry = map.get(keyword)!;
    if (category !== 'unknown' && entry.category !== category) {
      return undefined;
    }
    return entry;
  }

  public static peek(range: TextRange) {
    const search = range.toString();
    const result: ParsedResult = {};
    const predefined: TextRange[] = [];
    for (const { category, list } of this.peekEntries) {
      for (const key of list) {
        // StringComparison.CurrentCulture
        // https://learn.microsoft.com/ja-jp/dotnet/api/system.stringcomparison?view=net-7.0
        const foundIdx = search.indexOf(key);
        if (foundIdx === -1) continue;
        result[category] = key;
        predefined.push(range.fork(foundIdx + range.offset, key.length));
      }
    }
    return { result, predefined };
  }
}

export const Fansubs = new Set([
  '猎户发布组',
  '猎户手抄部',
  '北宇治字幕组',
  '北宇治Anarchism字幕组',
  '動漫花園',
  '拨雪寻春',
  'NC-Raws',
  '喵萌奶茶屋',
  'Lilith-Raws',
  '魔星字幕团',
  '桜都字幕组',
  '天月動漫&發佈組',
  '极影字幕社',
  'LoliHouse',
  '悠哈C9字幕社',
  '幻月字幕组',
  '天使动漫论坛',
  '动漫国字幕组',
  '幻樱字幕组',
  '爱恋字幕社',
  'DBD制作组',
  'c.c动漫',
  '萝莉社活动室',
  '千夏字幕组',
  'IET字幕組',
  '诸神kamigami字幕组',
  '霜庭云花Sub',
  'GMTeam',
  '风车字幕组',
  // '雪飄工作室(FLsnow)',
  'MCE汉化组',
  '丸子家族',
  '星空字幕组',
  '梦蓝字幕组',
  'LoveEcho!',
  'SweetSub',
  '枫叶字幕组',
  'Little Subbers!',
  '轻之国度',
  '云光字幕组',
  '豌豆字幕组',
  '驯兽师联盟',
  '中肯字幕組',
  'SW字幕组',
  '银色子弹字幕组',
  '风之圣殿',
  'YWCN字幕组',
  'KRL字幕组',
  '华盟字幕社',
  '波洛咖啡厅',
  '动音漫影',
  'VCB-Studio',
  'DHR動研字幕組',
  '80v08',
  '肥猫压制',
  'Little字幕组',
  'AI-Raws',
  '离谱Sub',
  '虹咲学园烤肉同好会',
  'ARIA吧汉化组',
  '柯南事务所',
  '百冬練習組',
  '冷番补完字幕组',
  '爱咕字幕组',
  '極彩字幕组',
  'AQUA工作室',
  '未央阁联盟',
  '届恋字幕组',
  '夜莺家族',
  'TD-RAWS',
  '夢幻戀櫻',
  'WBX-SUB',
  'Liella!の烧烤摊',
  'Amor字幕组',
  'MingYSub',
  '小白GM',
  'Sakura',
  'EMe',
  'Alchemist',
  '黑岩射手吧字幕组',
  'ANi',
  'MSB制作組'
]);
