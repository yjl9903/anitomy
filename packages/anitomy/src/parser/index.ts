import type { AnitomyOptions, ParsedResult } from '../types';

import { KeywordManager } from '../keyword';
import { ElementCategory } from '../element';
import { Token, TokenCategory, TokenFlag, findPrevToken, findToken } from '../token';
import { inRange, isNumericString, trim } from '../utils';

import {
  isCRC32,
  isElementCategorySearchable,
  isElementCategorySingular,
  isMatchTokenCategory,
  isMostlyLatinString,
  isResolution
} from './utils';
import {
  AnimeYearMax,
  AnimeYearMin,
  indexOfDigit,
  searchForLastNumber,
  searchForSeparatedNumbers,
  searchForEquivalentNumbers,
  searchForIsolatedEpisodeNumber,
  searchForEpisodeNumberWithVersion
} from './number';
import { ParserContext, hasResult, setResult } from './context';
import {
  checkAndSetAnimeSeasonKeyword,
  checkAndSetAnimeSeason,
  checkExtentKeyword,
  isTokenIsolated,
  buildElement,
  checkAndSetAnimeMonth,
  checkAndSetReleaseGroup
} from './parser';
import { searchForEpisodePatterns } from './episode';

export function parse(result: ParsedResult, tokens: Token[], options: AnitomyOptions) {
  const context: ParserContext = {
    tokens,
    options,
    result,
    isEpisodeKeywordsFound: false
  };

  searchForKeywords(context);
  searchForIsolatedNumbers(context);

  if (options.parseEpisodeNumber) {
    searchForEpisodeNumber(context);
  }

  searchForAnimeTitle(context);

  if (options.parseReleaseGroup && !hasResult(context, ElementCategory.ReleaseGroup)) {
    searchForReleaseGroup(context);
  }

  if (options.parseEpisodeTitle && hasResult(context, ElementCategory.EpisodeNumber)) {
    searchForEpisodeTitle(context);
  }

  validateElements(context);

  return {
    ok: hasResult(context, ElementCategory.AnimeTitle),
    result: context.result
  };
}

function searchForKeywords(context: ParserContext) {
  const tokens = context.tokens;
  const options = context.options;

  for (let i = 0; i < tokens.length; i++) {
    const token = tokens[i];
    if (token.category !== TokenCategory.Unknown) continue;

    let word = trim(token.content, [' ', '-']);
    if (word === '') continue;

    // Don't bother if the word is a number that cannot be CRC
    if (word.length !== 8 && isNumericString(word)) continue;

    let category = ElementCategory.Unknown;
    const keyword = KeywordManager.normalize(word);
    const found = KeywordManager.find(keyword, ElementCategory.Unknown);
    if (found) {
      category = found.category;
      if (!options.parseReleaseGroup && category === ElementCategory.ReleaseGroup) {
        continue;
      }
      if (!isElementCategorySearchable(category) || !found.searchable) {
        continue;
      }
      if (isElementCategorySingular(category) && hasResult(context, category)) {
        continue;
      }

      switch (found.category) {
        case ElementCategory.AnimeSeasonPrefix:
          checkAndSetAnimeSeasonKeyword(context, i);
          continue;
        case ElementCategory.EpisodePrefix:
          if (found.valid) {
            checkExtentKeyword(context, ElementCategory.EpisodeNumber, i);
          }
          continue;
        case ElementCategory.ReleaseVersion:
          word = word.slice(1);
          break;
        case ElementCategory.VolumePrefix:
          checkExtentKeyword(context, ElementCategory.VolumeNumber, i);
          continue;
      }
    } else {
      // Added by https://github.com/yjl9903/AnimeGarden
      // e.g. S2, S02, 第2季, 第二季
      if (checkAndSetAnimeSeason(context, i)) {
        continue;
      }

      // Added by https://github.com/yjl9903/AnimeGarden
      // e.g. 1月新番, 4月新番
      if (checkAndSetAnimeMonth(context, i)) {
        continue;
      }

      // Added by https://github.com/yjl9903/AnimeGarden
      if (checkAndSetReleaseGroup(context, i)) {
        continue;
      }

      // CRC32 checksum
      if (!hasResult(context, ElementCategory.FileChecksum) && isCRC32(word)) {
        category = ElementCategory.FileChecksum;
      } else if (!hasResult(context, ElementCategory.VideoResolution) && isResolution(word)) {
        category = ElementCategory.VideoResolution;
      }
      // Added by https://github.com/yjl9903/AnimeGarden
      // Video extension may appear in the title
      {
        const found = KeywordManager.find(keyword, ElementCategory.FileExtension);
        if (found) {
          category = found.category;
        }
      }
    }

    if (category !== ElementCategory.Unknown) {
      setResult(context, category, word);
      if (!found || found.identifiable) {
        token.category = TokenCategory.Identifier;
      }
    }
  }
}

function searchForIsolatedNumbers(context: ParserContext) {
  for (let i = 0; i < context.tokens.length; i++) {
    const token = context.tokens[i];
    if (
      token.category !== TokenCategory.Unknown ||
      !isNumericString(token.content) ||
      !isTokenIsolated(context, i)
    ) {
      continue;
    }

    // Anime Year
    const num = +token.content;
    if (AnimeYearMin <= num && num <= AnimeYearMax) {
      if (!hasResult(context, ElementCategory.AnimeYear)) {
        setResult(context, ElementCategory.AnimeYear, token.content);
        token.category = TokenCategory.Identifier;
        continue;
      }
    }

    // Video Resolution
    if (num !== 480 && num !== 720 && num !== 1080) continue;
    // If these numbers are isolated, it's more likely for them to be the
    // video resolution rather than the episode number. Some fansub groups use these without the "p" suffix.
    if (hasResult(context, ElementCategory.VideoResolution)) continue;
    setResult(context, ElementCategory.VideoResolution, token.content);
    token.category = TokenCategory.Identifier;
  }
}

function searchForEpisodeNumber(context: ParserContext) {
  const tokens = context.tokens
    .map((token, idx) => [token, idx] as const)
    .filter(
      // fix: only use Unknown token and it must have digit char
      ([token]) => token.category === TokenCategory.Unknown && indexOfDigit(token.content) !== -1
    )
    .map(([_token, idx]) => idx);
  if (tokens.length === 0) return;

  context.isEpisodeKeywordsFound = hasResult(context, ElementCategory.EpisodeNumber);

  // If a token matches a known episode pattern, it has to be the episode number
  if (searchForEpisodePatterns(context, tokens)) return;

  // We have previously found an episode number via keywords
  context.isEpisodeKeywordsFound = hasResult(context, ElementCategory.EpisodeNumber);
  if (context.isEpisodeKeywordsFound) return;

  // From now on, we're only interested in numeric tokens
  tokens.splice(
    0,
    tokens.length,
    ...tokens.filter((t) => isNumericString(context.tokens[t].content))
  );

  // e.g. "01 (176)", "29 (04)"
  if (searchForEquivalentNumbers(context, tokens)) return;

  // e.g. " - 08"
  if (searchForSeparatedNumbers(context, tokens)) return;

  // e.g. "[12]", "(2006)"
  if (searchForIsolatedEpisodeNumber(context, tokens)) return;

  // e.g. " 03 v2"
  if (searchForEpisodeNumberWithVersion(context, tokens)) return;

  // Consider using the last number as a last resort
  searchForLastNumber(context, tokens);
}

function searchForAnimeTitle(context: ParserContext) {
  let enclosedTitle = false;
  let tokenBegin = findToken(
    context.tokens,
    0,
    context.tokens.length,
    TokenFlag.NotEnclosed,
    TokenFlag.Unknown
  );

  if (!inRange(context.tokens, tokenBegin)) {
    tokenBegin = 0;
    enclosedTitle = true;
    do {
      tokenBegin = findToken(context.tokens, tokenBegin, context.tokens.length, TokenFlag.Unknown);
      if (!inRange(context.tokens, tokenBegin)) break;

      // Changed by https://github.com/yjl9903/AnimeGarden
      // Ignore groups that are composed of Latin characters
      if (!isMostlyLatinString(context.tokens[tokenBegin].content)) {
        break;
      }

      tokenBegin = findToken(context.tokens, tokenBegin, context.tokens.length, TokenFlag.Bracket);
      tokenBegin = findToken(context.tokens, tokenBegin, context.tokens.length, TokenFlag.Unknown);
    } while (inRange(context.tokens, tokenBegin));
  }

  if (!inRange(context.tokens, tokenBegin)) return;

  let tokenEnd = findToken(
    context.tokens,
    tokenBegin,
    context.tokens.length,
    TokenFlag.Identifier,
    enclosedTitle ? TokenFlag.Bracket : TokenFlag.None
  );

  // If within the interval there's an open bracket without its matching pair,
  // move the upper endpoint back to the bracket
  if (!enclosedTitle) {
    let lastBracket = tokenEnd;
    let bracketOpen = false;
    for (let i = tokenBegin; i < tokenEnd; i++) {
      if (context.tokens[i].category === TokenCategory.Bracket) {
        lastBracket = i;
        bracketOpen = !bracketOpen;
      }
    }
    if (bracketOpen) tokenEnd = lastBracket;
  }

  // If the interval ends with an enclosed group (e.g. "Anime Title [Fansub]"),
  // move the upper endpoint back to the beginning of the group. We ignore
  // parentheses in order to keep certain groups (e.g. "(TV)") intact.
  if (!enclosedTitle) {
    let token = findPrevToken(context.tokens, tokenEnd, TokenFlag.NotDelimiter);
    while (
      isMatchTokenCategory(TokenCategory.Bracket, context.tokens[token]) &&
      context.tokens[token].content[0] != ')'
    ) {
      token = findPrevToken(context.tokens, token, TokenFlag.Bracket);
      if (inRange(context.tokens, token)) {
        tokenEnd = token;
        token = findPrevToken(context.tokens, tokenEnd, TokenFlag.NotDelimiter);
      }
    }
  }

  // Build title element
  buildElement(
    context,
    ElementCategory.AnimeTitle,
    false,
    context.tokens.slice(tokenBegin, tokenEnd)
  );
}

function searchForReleaseGroup(context: ParserContext) {}

function searchForEpisodeTitle(context: ParserContext) {}

function validateElements(context: ParserContext) {}
