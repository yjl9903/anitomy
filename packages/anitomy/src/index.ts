import type { AnitomyOptions, AnitomyResult, ParsedResult } from './types';

import { mergeResult } from './utils';
import { KeywordManager } from './keyword';
import { ElementCategory } from './element';
import { parse as doParse } from './parser';
import { tokenize as doTokenize } from './tokenizer';

export type * from './types';

/**
 * Parser instance with shared options and parsed result cache
 */
export class Parser {
  private readonly options: AnitomyOptions;

  private readonly cache = new Map<string, AnitomyResult | null>();

  public constructor(options: Partial<AnitomyOptions> = {}) {
    this.options = resolveOptions(options);
  }

  public parse(filename: string, force = false) {
    if (!force && this.cache.has(filename)) {
      const result = this.cache.get(filename);
      return result ? result : undefined;
    } else {
      const result = parse(filename, this.options);
      if (result) {
        this.cache.set(filename, result);
      } else {
        this.cache.set(filename, null);
      }
      return result;
    }
  }
}

export function parse(
  filename: string,
  _options: Partial<AnitomyOptions> = {}
): AnitomyResult | undefined {
  if (filename === '') return undefined;

  let result: ParsedResult = {};

  const options = resolveOptions(_options);

  result.filename = filename;
  if (options.parseFileExtension) {
    const ext = removeExtension(filename);
    if (ext) {
      result.filename = filename;
      result.extension = ext.extension;
    }
  }

  const tokenized = doTokenize(result.filename!, options);
  result = mergeResult(result, tokenized.result);
  if (!tokenized.ok) {
    return resolveResult(result);
  }

  const parsed = doParse(result, tokenized.tokens, options);
  result = parsed.result;

  return resolveResult(result);
}

export function resolveOptions(options: Partial<AnitomyOptions>): AnitomyOptions {
  return {
    delimiters: ' _.&+,|',
    parseEpisodeNumber: true,
    parseEpisodeTitle: true,
    parseFileExtension: true,
    parseReleaseGroup: true,
    ...options
  };
}

function resolveResult(result: ParsedResult): AnitomyResult {
  const resolved: AnitomyResult = {
    title: result['title'],
    type: result['type'],
    season: normalizeSeason(result['season']),
    year: normalizeNumber(result['year']),
    month: normalizeNumber(result['month']),
    language: result['language'],
    subtitles: result['subtitles'],
    source: result['source'],
    episode: {
      number: normalizeNumber(result['episode.number']),
      numberAlt: normalizeNumber(result['episode.numberAlt']),
      title: result['episode.title']
    },
    volume: {
      number: normalizeNumber(result['volume'])
    },
    video: {
      term: result['video.term'],
      resolution: result['video.resolution']
    },
    audio: {
      term: result['audio.term']
    },
    release: {
      version: normalizeNumber(result['release.version']),
      group: result['release.group']
    },
    file: {
      name: result['filename']!,
      extension: result['extension'],
      checksum: result['checksum']
    }
  };
  return resolved;

  function normalizeSeason(num: string | undefined) {
    if (num !== undefined && num !== null) {
      return /^\d+$/.test(num) ? String(+num) : num;
    } else {
      return undefined;
    }
  }

  function normalizeNumber(num: string | undefined) {
    try {
      if (num !== undefined && num !== null) {
        const n = parseFloat(num);
        return !Number.isNaN(n) ? n : undefined;
      }
    } catch {}
    return undefined;
  }
}

function removeExtension(filename: string) {
  const position = filename.lastIndexOf('.');
  if (position === -1) return undefined;

  const extension = filename.slice(position + 1);
  if (extension.length > 4 || !/^[a-zA-Z0-0]+$/.test(extension)) {
    return undefined;
  }

  if (!KeywordManager.contains(ElementCategory.FileExtension, extension)) {
    return undefined;
  }

  return {
    filename: filename.slice(0, position),
    extension
  };
}
