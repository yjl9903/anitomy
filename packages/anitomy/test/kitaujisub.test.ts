import { describe, it, expect } from 'vitest';

import { parse } from '../src';

import { KitaujiSub } from './testcase';

describe('Kitaujisub', () => {
  it('should parse', () => {
    for (const filename of KitaujiSub) {
      const info = parse(filename);
      expect(info).toMatchSnapshot();
      expect(info?.title).toBeTruthy();
    }
  });
});
