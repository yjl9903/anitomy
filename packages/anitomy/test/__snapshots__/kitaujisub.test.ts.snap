// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Kitaujisub > should parse 1`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": undefined,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[北宇治字幕组&7³ACG] 镜之孤城 / Kagami no Kojou [Webrip][1080p][HEVC_AAC][简繁内封]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "北宇治字幕组",
    "version": undefined,
  },
  "season": undefined,
  "source": "Webrip",
  "subtitles": undefined,
  "title": "镜之孤城 / Kagami no Kojou",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Kitaujisub > should parse 2`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": undefined,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "mkv",
    "name": "[KitaujiSub&7鲁ACG] Kagami no Kojou [Webrip][1080p][HEVC_AAC][CHS&CHT].mkv",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": undefined,
    "version": undefined,
  },
  "season": undefined,
  "source": "Webrip",
  "subtitles": undefined,
  "title": "Kagami no Kojou",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Kitaujisub > should parse 3`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 3,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[北宇治字幕组] 赛马娘 Pretty Derby Road to the Top / Uma Musume - Pretty Derby - Road to the Top [03 v2][Webrip][1080p][HEVC_AAC][简体内封]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "北宇治字幕组",
    "version": 2,
  },
  "season": undefined,
  "source": "Webrip",
  "subtitles": undefined,
  "title": "赛马娘 Pretty Derby Road to the Top / Uma Musume - Pretty Derby - Road to the Top",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Kitaujisub > should parse 4`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 3,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "mkv",
    "name": "[KitaujiSub] Uma Musume - Pretty Derby - Road to the Top [03][Webrip][1080P][HEVC_AAC][CHS][v2].mkv",
  },
  "language": "CHS",
  "month": undefined,
  "release": {
    "group": undefined,
    "version": 2,
  },
  "season": undefined,
  "source": "Webrip",
  "subtitles": undefined,
  "title": "Uma Musume - Pretty Derby - Road to the Top",
  "type": undefined,
  "video": {
    "resolution": "1080P",
    "term": "HEVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Kitaujisub > should parse 5`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 4,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[北宇治Anarchism字幕组] 魔法少女毀滅者/Mahou Shoujo Magical Destroyers [04][Webrip][1080p][HEVC_AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "北宇治Anarchism字幕组",
    "version": undefined,
  },
  "season": undefined,
  "source": "Webrip",
  "subtitles": undefined,
  "title": "魔法少女毀滅者/Mahou Shoujo Magical Destroyers",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Kitaujisub > should parse 6`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 3,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[北宇治字幕组&霜庭云花Sub&氢气烤肉架]【我推的孩子】/【Oshi no ko】[03 V2][Webrip][1080p][HEVC_AAC][简繁日内封]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "霜庭云花Sub",
    "version": 2,
  },
  "season": undefined,
  "source": "Webrip",
  "subtitles": undefined,
  "title": "【我推的孩子】/",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;
